<h1> City of Toronto Neighborhood Analysis </h1>

<h1>*********************************Part 1**************************</h1>

In [1]:
#from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page
import pandas as pd
import numpy as np

<h2> First task is to webscrap the data from wikipedia</h2>

In [2]:
#wikipedia page with Toronto Neighborhood Data
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' 

In [4]:
#Get the Requested page and convert it to text
data  = requests.get(url).text 

In [5]:
#Using Beautibul Soup to get the list of all dataframes in the requested website
dataframe_list = pd.read_html(url, flavor='bs4')
len(dataframe_list)

3

In [6]:
#checkink the first DataFrame out of 3
dataframe_list[0]

0  \
0                                     M1ANot assigned   
1                     M1BScarborough(Malvern / Rouge)   
2   M1CScarborough(Rouge Hill / Port Union / Highl...   
3   M1EScarborough(Guildwood / Morningside / West ...   
4                              M1GScarborough(Woburn)   
5                           M1HScarborough(Cedarbrae)   
6                 M1JScarborough(Scarborough Village)   
7   M1KScarborough(Kennedy Park / Ionview / East B...   
8   M1LScarborough(Golden Mile / Clairlea / Oakridge)   
9   M1MScarborough(Cliffside / Cliffcrest / Scarbo...   
10       M1NScarborough(Birch Cliff / Cliffside West)   
11  M1PScarborough(Dorset Park / Wexford Heights /...   
12                 M1RScarborough(Wexford / Maryvale)   
13                          M1SScarborough(Agincourt)   
14  M1TScarborough(Clarks Corners / Tam O'Shanter ...   
15  M1VScarborough(Milliken / Agincourt North / St...   
16     M1WScarborough(Steeles West / L'Amoreaux West)   
17                        M1XScarborough(Upper Rouge)   
18                                    M1YNot assigned   
19                                    M1ZNot assigned   

                                                1  \
0                                 M2ANot assigned   
1                                 M2BNot assigned   
2                                 M2CNot assigned   
3                                 M2ENot assigned   
4                                 M2GNot assigned   
5                M2HNorth York(Hillcrest Village)   
6   M2JNorth York(Fairview / Henry Farm / Oriole)   
7                  M2KNorth York(Bayview Village)   
8        M2LNorth York(York Mills / Silver Hills)   
9         M2MNorth York(Willowdale / Newtonbrook)   
10                 M2NNorth York(Willowdale)South   
11                 M2PNorth York(York Mills West)   
12                  M2RNorth York(Willowdale)West   
13                                M2SNot assigned   
14                                M2TNot assigned   
15                                M2VNot assigned   
16                                M2WNot assigned   
17                                M2XNot assigned   
18                                M2YNot assigned   
19                                M2ZNot assigned   

                                                    2  \
0                            M3ANorth York(Parkwoods)   
1                       M3BNorth York(Don Mills)North   
2      M3CNorth York(Don Mills)South(Flemingdon Park)   
3                                     M3ENot assigned   
4                                     M3GNot assigned   
5   M3HNorth York(Bathurst Manor / Wilson Heights ...   
6     M3JNorth York(Northwood Park / York University)   
7          M3KNorth York(Downsview)East (CFB Toronto)   
8                        M3LNorth York(Downsview)West   
9                     M3MNorth York(Downsview)Central   
10                  M3NNorth York(Downsview)Northwest   
11                                    M3PNot assigned   
12                                    M3RNot assigned   
13                                    M3SNot assigned   
14                                    M3TNot assigned   
15                                    M3VNot assigned   
16                                    M3WNot assigned   
17                                    M3XNot assigned   
18                                    M3YNot assigned   
19                                    M3ZNot assigned   

                                                    3  \
0                     M4ANorth York(Victoria Village)   
1      M4BEast York(Parkview Hill / Woodbine Gardens)   
2                      M4CEast York(Woodbine Heights)   
3                        M4EEast Toronto(The Beaches)   
4                               M4GEast York(Leaside)   
5                      M4HEast York(Thorncliffe Park)   
6         M4JEast YorkEast Toronto(The Danforth East)   
7      M4KEast Toronto(The Danforth West / Riverdale)   
8    M4LEast Toronto(India Ba

<h4>The first dataframe seems to contain the required data. 
Each element of the dataframe contains a string which has first 3 character as the postal code, next the Borough Name and next inside the bracket are the Neighbourhood names separated with '/' </h4>

In [7]:
dataframe_list[0].shape


(20, 9)

<h3>First we will create a dataFrame with 2 columns one with postal code and other with rest all Informations named RestInfo.</h3>
<h4>Later we will seperate the Borough and Neighborhood names from the RestInfo column</h4>

In [8]:
df = pd.DataFrame(columns=['PostalCode','RestInfo']) #Creating the dataframe with 2 columns
df_temp = dataframe_list[0]

In [9]:
#Iterating through each elements and adding the corresponding PostalCode and RestInfo in our DataFrame df
for rows in range(df_temp.shape[0]):
        for columns in range(df_temp.shape[1]):
            key=df_temp[columns][rows].lstrip()[0:3] #using the lstrip() func to ensure no blank space is counted. Then sing the 1st 3 character as postalcode
            val=df_temp[columns][rows].lstrip()[3:] #assigning rest of the string as RestInfo
            new_row = {'PostalCode':key,'RestInfo':val}
            df = df.append(new_row, ignore_index=True)
df

PostalCode                                           RestInfo
0          M1A                                       Not assigned
1          M2A                                       Not assigned
2          M3A                              North York(Parkwoods)
3          M4A                       North York(Victoria Village)
4          M5A       Downtown Toronto(Regent Park / Harbourfront)
..         ...                                                ...
175        M5Z                                       Not assigned
176        M6Z                                       Not assigned
177        M7Z                                       Not assigned
178        M8Z  Etobicoke(Mimico NW / The Queensway West / Sou...
179        M9Z                                       Not assigned

[180 rows x 2 columns]

In [10]:
# in RestInfo Column anything before the bracket '(' is Borough and Info inside the bracket are neigborhood
# So we split the RestInfo in two part (seperated by "(" ), the first part is Borough and Second id Neighborhoods
# We use two separate functions to get the result
def find_Borough(x):
    if '(' in x:
        return x.split('(')[0]
    else:
        return x
    
    
def find_Neighb(x):
    if '(' in x:
        str_len = len(x.split('(')[1])
        return x.split('(')[1][:str_len-1] # skipping the last character as it would be the closing bracket ')'.
    else:
        return 

In [11]:
#use the functions defines above to seperate Borough and Neighborhood
df['Borough'] = df['RestInfo'].apply(find_Borough)
df['Neighbourhood'] = df['RestInfo'].apply(find_Neighb)
df.head()

PostalCode                                      RestInfo           Borough  \
0        M1A                                  Not assigned      Not assigned   
1        M2A                                  Not assigned      Not assigned   
2        M3A                         North York(Parkwoods)        North York   
3        M4A                  North York(Victoria Village)        North York   
4        M5A  Downtown Toronto(Regent Park / Harbourfront)  Downtown Toronto   

                Neighbourhood  
0                        None  
1                        None  
2                   Parkwoods  
3            Victoria Village  
4  Regent Park / Harbourfront

In [12]:
#dropping the RestInfo column
df.drop(columns=['RestInfo'],axis=1, inplace=True)
df.tail()

PostalCode       Borough  \
175        M5Z  Not assigned   
176        M6Z  Not assigned   
177        M7Z  Not assigned   
178        M8Z     Etobicoke   
179        M9Z  Not assigned   

                                         Neighbourhood  
175                                               None  
176                                               None  
177                                               None  
178  Mimico NW / The Queensway West / South of Bloo...  
179                                               None

In [13]:
#replace the slash '/' with ',' in Neighboorhood columns
nl=None
df['Neighbourhood'] = df['Neighbourhood'].apply(lambda x: None if x==nl else x.replace('/',','))
df['Neighbourhood']

0                                                   None
1                                                   None
2                                              Parkwoods
3                                       Victoria Village
4                             Regent Park , Harbourfront
                             ...                        
175                                                 None
176                                                 None
177                                                 None
178    Mimico NW , The Queensway West , South of Bloo...
179                                                 None
Name: Neighbourhood, Length: 180, dtype: object

<h3>As per the Problem Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.</h3>
<H2>Now we need to delete the rows where Borough is Not assigned.</H2>

In [14]:
df.drop(df[df['Borough']=='Not assigned'].index,inplace =True)
df.reset_index(drop=True, inplace =True)
df

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                           Regent Park , Harbourfront  
3                    Lawrence Manor , Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98     The Kingsway , Montgomery Road , Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South , King's Mill Park , Sunnylea ,...  
102  Mimico NW , The Queensway West , South of Bloo...  

[103 rows x 3 columns]

<h2> Check if any Postal Code is listed twice </h2>

In [15]:
len(df['PostalCode'].unique())

103

<h4>Since there are 103 unique value of Postal Code and there are 103 rows in the table, there seems no double entry of postal code</h4>

<h3>According to the problem If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.
So we check: </h3>
<h2> If there is a Not Assigned Neighborhood</h2>

In [16]:
df[df['Neighbourhood'].isnull()]

Empty DataFrame
Columns: [PostalCode, Borough, Neighbourhood]
Index: []

In [17]:
df[df['Neighbourhood']==" "]

Empty DataFrame
Columns: [PostalCode, Borough, Neighbourhood]
Index: []

In [18]:
df[df['Neighbourhood']== np.nan]

Empty DataFrame
Columns: [PostalCode, Borough, Neighbourhood]
Index: []

<h4> so according to the above tests all Neighbourhood is assigned to some value and there is no Neighborhood with null value or Not assigned value. Therefor no further operation need to be done here </h4>

In [19]:
df.head()

PostalCode           Borough                      Neighbourhood
0        M3A        North York                          Parkwoods
1        M4A        North York                   Victoria Village
2        M5A  Downtown Toronto         Regent Park , Harbourfront
3        M6A        North York  Lawrence Manor , Lawrence Heights
4        M7A      Queen's Park      Ontario Provincial Government

In [20]:
df.shape

(103, 3)

In [21]:
df.to_csv('TorontoPostalCode.csv',index=False) #saving the file so that the whole part1 need not to be run again for part 2

<h1>***************************Part 2 *********************************</h1>

In [1]:
#import geocoder # import geocoder
!conda install -c conda-forge geopy --yes


Solving environment: ...working... done

# All requested packages already installed.



In [2]:
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: ...working... done

# All requested packages already installed.



In [3]:
from geopy.geocoders import Nominatim
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library


In [4]:
df1=pd.read_csv('TorontoPostalCode.csv')
df1.head()

PostalCode           Borough                      Neighbourhood
0        M3A        North York                          Parkwoods
1        M4A        North York                   Victoria Village
2        M5A  Downtown Toronto         Regent Park , Harbourfront
3        M6A        North York  Lawrence Manor , Lawrence Heights
4        M7A      Queen's Park      Ontario Provincial Government

In [5]:
#Function to get latitude and Longitude using geolocator
def get_coordinates(postal_code):
    geolocator = Nominatim(user_agent="foursquare_agent")
    g = geolocator.geocode('{}, Toronto, Ontario'.format(postal_code))
    try:
        latitude = g.latitude
        longitude = g.longitude
    except:
        return None
    return latitude,longitude

In [7]:
#checking if the function works correctly
get_coordinates('M7A')

(43.6534817, -79.3839347)

In [8]:
df1['Latitude_Longitude']= df1['PostalCode'].apply(get_coordinates)

In [9]:
df1['Latitude_Longitude']

0                    (43.6534817, -79.3839347)
1                                         None
2                                         None
3                                         None
4                    (43.6534817, -79.3839347)
5                    (43.6534817, -79.3839347)
6                    (43.6534817, -79.3839347)
7                                         None
8                                         None
9                                         None
10                                        None
11                                        None
12                   (43.6534817, -79.3839347)
13                                        None
14                                        None
15                                        None
16                                        None
17     (43.64410993066079, -79.58890692194828)
18                                        None
19                                        None
20                   (43.6421064, -79.3774455)
21           

In [10]:
len(df1['Latitude_Longitude'].unique())

18

coordinates of only 18 postalArea could be found 

In [11]:
df1.drop(columns=['Latitude_Longitude'], inplace = True)

<h4> Due to time out error and inefficiency of the above Geolocator service we will skip using geolocator and use the Geospatial_Coordinates.csv as provided to find the latitude and longitude</h4>

In [12]:
lat_lng_file = pd.read_csv('Geospatial_Coordinates.csv')
df2=pd.merge(df1, lat_lng_file, how='inner', left_on='PostalCode', right_on='Postal Code',left_index=False, right_index=False)
df2.head()

PostalCode           Borough                      Neighbourhood Postal Code  \
0        M3A        North York                          Parkwoods         M3A   
1        M4A        North York                   Victoria Village         M4A   
2        M5A  Downtown Toronto         Regent Park , Harbourfront         M5A   
3        M6A        North York  Lawrence Manor , Lawrence Heights         M6A   
4        M7A      Queen's Park      Ontario Provincial Government         M7A   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [13]:
df2.drop(columns=['Postal Code'],inplace=True)
df2.head()

PostalCode           Borough                      Neighbourhood   Latitude  \
0        M3A        North York                          Parkwoods  43.753259   
1        M4A        North York                   Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park , Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor , Lawrence Heights  43.718518   
4        M7A      Queen's Park      Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [14]:
df2.shape

(103, 5)

<h1>******************************Part 3*********************************</h1>

In [15]:
import requests 

<h4> Finding all the Borough with name 'Toronto' in it</h4>

In [16]:

df3 = df2[df2['Borough'].apply(lambda x: True if 'Toronto' in x else False)]
df3.reset_index(drop=True)
df3

PostalCode                                            Borough  \
2          M5A                                   Downtown Toronto   
9          M5B                                   Downtown Toronto   
15         M5C                                   Downtown Toronto   
19         M4E                                       East Toronto   
20         M5E                                   Downtown Toronto   
24         M5G                                   Downtown Toronto   
25         M6G                                   Downtown Toronto   
30         M5H                                   Downtown Toronto   
31         M6H                                       West Toronto   
35         M4J                              East YorkEast Toronto   
36         M5J                                   Downtown Toronto   
37         M6J                                       West Toronto   
41         M4K                                       East Toronto   
42         M5K                                   Downtown Toronto   
43         M6K                                       West Toronto   
47         M4L                                       East Toronto   
48         M5L                                   Downtown Toronto   
54         M4M                                       East Toronto   
61         M4N                                    Central Toronto   
62         M5N                                    Central Toronto   
67         M4P                                    Central Toronto   
68         M5P                                    Central Toronto   
69         M6P                                       West Toronto   
73         M4R                                    Central Toronto   
74         M5R                                    Central Toronto   
75         M6R                                       West Toronto   
79         M4S                                    Central Toronto   
80         M5S                                   Downtown Toronto   
81         M6S                                       West Toronto   
83         M4T                                    Central Toronto   
84         M5T                                   Downtown Toronto   
86         M4V                                    Central Toronto   
87         M5V                                   Downtown Toronto   
91         M4W                                   Downtown Toronto   
92         M5W     Downtown TorontoStn A PO Boxes25 The Esplanade   
96         M4X                                   Downtown Toronto   
97         M5X                                   Downtown Toronto   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   

                                         Neighbourhood   Latitude  Longitude  
2                           Regent Park , Harbourfront  43.654260 -79.360636  
9                             Garden District, Ryerson  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                          Richmond , Adelaide , King  43.650571 -79.384568  
31                       Dufferin , Dovercourt Village  43.669005 -79.442259  
35                                   The Danforth East  43.685347 -79.338106  
36   Harbourfront East , Union Station , Toronto Is...  43.640816 -79.381752  
37                           Little Portugal , Trinity  43.647927 -79.419750  
41                       The Danforth West , Riverdale  43.679557 -79.352188  
42           Toronto Dominion Centre , Design Exchange  43.647177 -79.381576  
43      Brockton , Parkdale Village , Exhibition Place  43

In [86]:
len(df3['Borough'].unique())

7

<h4> Now we prepare the Credentials to access Foursquare API</h4>

In [18]:
CLIENT_ID = 'EDVNUDPMPCTF4YZYEHJ5HF0UQEX5LLKI2RZ5NT21ELCZ5DON' # your Foursquare ID
CLIENT_SECRET = 'HW5DJH3G5BSDXF1JAMDLI5SIQZZIFDIMTPOBQNOACQ3CWHLT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: EDVNUDPMPCTF4YZYEHJ5HF0UQEX5LLKI2RZ5NT21ELCZ5DON
CLIENT_SECRET:HW5DJH3G5BSDXF1JAMDLI5SIQZZIFDIMTPOBQNOACQ3CWHLT


<h4>Now, let's get the top 100 venues that are in area with Postal Code 'M5A' within a radius of 500 meters.</h4>

In [19]:
neighborhood_latitude=df3[(df3['PostalCode']=='M5A').values]['Latitude'].values[0]
neighborhood_longitude=df3[(df3['PostalCode']=='M5A').values]['Longitude'].values[0]

In [20]:
neighborhood_latitude,neighborhood_longitude

(43.6542599, -79.3606359)

Preparing the url

In [21]:

LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=EDVNUDPMPCTF4YZYEHJ5HF0UQEX5LLKI2RZ5NT21ELCZ5DON&client_secret=HW5DJH3G5BSDXF1JAMDLI5SIQZZIFDIMTPOBQNOACQ3CWHLT&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

Requesting the Resultant JSON file

In [22]:
results = requests.get(url).json()
json_formatted_str = json.dumps(results, indent=6)
print(json_formatted_str)

{
      "meta": {
            "code": 200,
            "requestId": "60deca4c28ef2f25c43e57fe"
      },
      "response": {
            "headerLocation": "Corktown",
            "headerFullLocation": "Corktown, Toronto",
            "headerLocationGranularity": "neighborhood",
            "totalResults": 42,
            "suggestedBounds": {
                  "ne": {
                        "lat": 43.6587599045,
                        "lng": -79.3544279001486
                  },
                  "sw": {
                        "lat": 43.6497598955,
                        "lng": -79.36684389985142
                  }
            },
            "groups": [
                  {
                        "type": "Recommended Places",
                        "name": "recommended",
                        "items": [
                              {
                                    "reasons": {
                                          "count": 0,
                                          "

In [23]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Converting the data in the JSON format to DataFrame 

In [24]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-24-561c05f0fdd1>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name           categories        lat        lng
0           Tandem Coffee          Coffee Shop  43.653559 -79.361809
1        Roselle Desserts               Bakery  43.653447 -79.362017
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4          Impact Kitchen           Restaurant  43.656369 -79.356980

<h4> Now we will repeat the same process (as done above for only one postal code 'M5A') for all postal Code </h4>

In [25]:
#definning a function to extract data from JSON to DataFrame for all Postal Code
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode', 
                  'Postal Latitude', 
                  'Postal Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [26]:
toronto_venues = getNearbyVenues(names=df3['PostalCode'],
                                   latitudes=df3['Latitude'],
                                   longitudes=df3['Longitude']
                                  )

M5A
M5B
M5C
M4E
M5E
M5G
M6G
M5H
M6H
M4J
M5J
M6J
M4K
M5K
M6K
M4L
M5L
M4M
M4N
M5N
M4P
M5P
M6P
M4R
M5R
M6R
M4S
M5S
M6S
M4T
M5T
M4V
M5V
M4W
M5W
M4X
M5X
M4Y
M7Y


In [27]:
# Checking the size of the DataFrame
print(toronto_venues.shape)
toronto_venues.head(15)

(1587, 7)


PostalCode  Postal Latitude  Postal Longitude  \
0         M5A         43.65426        -79.360636   
1         M5A         43.65426        -79.360636   
2         M5A         43.65426        -79.360636   
3         M5A         43.65426        -79.360636   
4         M5A         43.65426        -79.360636   
5         M5A         43.65426        -79.360636   
6         M5A         43.65426        -79.360636   
7         M5A         43.65426        -79.360636   
8         M5A         43.65426        -79.360636   
9         M5A         43.65426        -79.360636   
10        M5A         43.65426        -79.360636   
11        M5A         43.65426        -79.360636   
12        M5A         43.65426        -79.360636   
13        M5A         43.65426        -79.360636   
14        M5A         43.65426        -79.360636   

                                   Venue  Venue Latitude  Venue Longitude  \
0                          Tandem Coffee       43.653559       -79.361809   
1                       Roselle Desserts       43.653447       -79.362017   
2                 Cooper Koo Family YMCA       43.653249       -79.358008   
3                    Body Blitz Spa East       43.654735       -79.359874   
4                         Impact Kitchen       43.656369       -79.356980   
5       The Distillery Historic District       43.650244       -79.359323   
6               Distillery Sunday Market       43.650075       -79.361832   
7                    SOMA chocolatemaker       43.650622       -79.358127   
8                                   Arvo       43.649963       -79.361442   
9                        Sumach Espresso       43.658135       -79.359515   
10                        Rooster Coffee       43.651900       -79.365609   
11  Young Centre for the Performing Arts       43.650825       -79.357593   
12                        Underpass Park       43.655764       -79.354806   
13                            The Aviary       43.653634       -79.354662   
14                Parliament Square Park       43.650264       -79.362195   

           Venue Category  
0             Coffee Shop  
1                  Bakery  
2     Distribution Center  
3                     Spa  
4              Restaurant  
5           Historic Site  
6          Farmers Market  
7          Chocolate Shop  
8             Coffee Shop  
9             Coffee Shop  
10            Coffee Shop  
11  Performing Arts Venue  
12                   Park  
13                    Pub  
14                   Park

Check how many results are obtained for each Postal Code

In [28]:
toronto_venues.groupby('PostalCode').count()

Postal Latitude  Postal Longitude  Venue  Venue Latitude  \
PostalCode                                                             
M4E                       4                 4      4               4   
M4J                       2                 2      2               2   
M4K                      42                42     42              42   
M4L                      23                23     23              23   
M4M                      36                36     36              36   
M4N                       3                 3      3               3   
M4P                       7                 7      7               7   
M4R                      20                20     20              20   
M4S                      37                37     37              37   
M4T                       2                 2      2               2   
M4V                      13                13     13              13   
M4W                       4                 4      4               4   
M4X                      43                43     43              43   
M4Y                      79                79     79              79   
M5A                      42                42     42              42   
M5B                     100               100    100             100   
M5C                      81                81     81              81   
M5E                      58                58     58              58   
M5G                      65                65     65              65   
M5H                      98                98     98              98   
M5J                     100               100    100             100   
M5K                     100               100    100             100   
M5L                     100               100    100             100   
M5N                       3                 3      3               3   
M5P                       4                 4      4               4   
M5R                      21                21     21              21   
M5S                      33                33     33              33   
M5T                      66                66     66              66   
M5V                      16                16     16              16   
M5W                      99                99     99              99   
M5X                     100               100    100             100   
M6G                      16                16     16              16   
M6H                      14                14     14              14   
M6J                      42                42     42              42   
M6K                      24                24     24              24   
M6P                      25                25     25              25   
M6R                      14                14     14              14   
M6S                      33                33     33              33   
M7Y                      18                18     18              18   

            Venue Longitude  Venue Category  
PostalCode                                   
M4E                       4               4  
M4J                       2               2  
M4K                      42              42  
M4L                      23              23  
M4M                      36              36  
M4N                       3               3  
M4P                       7               7  
M4R                      20              20  
M4S                      37              37  
M4T                       2               2  
M4V                      13              13  
M4W                       4               4  
M4X                      43              43  
M4Y                      79              79  
M5A                      42              42  
M5B                     100             100  
M5C                      81              81  
M5E                      58              58  
M5G                      65              65  
M5H                      98              98  
M5J                     100             100  
M

In [30]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


Use one Hot Encoding on all categories for ease of applying ML algo like KNearest Neighbor

In [31]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['PostalCode'] = toronto_venues['PostalCode'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

PostalCode  Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0        M5A               0        0                   0               0   
1        M5A               0        0                   0               0   
2        M5A               0        0                   0               0   
3        M5A               0        0                   0               0   
4        M5A               0        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       1     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0                          0   
4              0         0        0     0                          0   

   

In [32]:
toronto_onehot.shape

(1587, 232)

#### Next, let's group rows by Postal Codes and by taking the mean of the frequency of occurrence of each category

In [33]:
toronto_grouped = toronto_onehot.groupby('PostalCode').mean().reset_index()
toronto_grouped

PostalCode  Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0         M4E        0.000000   0.0000              0.0000           0.000   
1         M4J        0.000000   0.0000              0.0000           0.000   
2         M4K        0.000000   0.0000              0.0000           0.000   
3         M4L        0.000000   0.0000              0.0000           0.000   
4         M4M        0.000000   0.0000              0.0000           0.000   
5         M4N        0.000000   0.0000              0.0000           0.000   
6         M4P        0.000000   0.0000              0.0000           0.000   
7         M4R        0.000000   0.0000              0.0000           0.000   
8         M4S        0.000000   0.0000              0.0000           0.000   
9         M4T        0.000000   0.0000              0.0000           0.000   
10        M4V        0.000000   0.0000              0.0000           0.000   
11        M4W        0.000000   0.0000              0.0000           0.000   
12        M4X        0.000000   0.0000              0.0000           0.000   
13        M4Y        0.012658   0.0000              0.0000           0.000   
14        M5A        0.000000   0.0000              0.0000           0.000   
15        M5B        0.000000   0.0000              0.0000           0.000   
16        M5C        0.000000   0.0000              0.0000           0.000   
17        M5E        0.000000   0.0000              0.0000           0.000   
18        M5G        0.000000   0.0000              0.0000           0.000   
19        M5H        0.000000   0.0000              0.0000           0.000   
20        M5J        0.000000   0.0000              0.0000           0.000   
21        M5K        0.000000   0.0000              0.0000           0.000   
22        M5L        0.000000   0.0000              0.0000           0.000   
23        M5N        0.000000   0.0000              0.0000           0.000   
24        M5P        0.000000   0.0000              0.0000           0.000   
25        M5R        0.000000   0.0000              0.0000           0.000   
26        M5S        0.000000   0.0000              0.0000           0.000   
27        M5T        0.000000   0.0000              0.0000           0.000   
28        M5V        0.000000   0.0625              0.0625           0.125   
29        M5W        0.000000   0.0000              0.0000           0.000   
30        M5X        0.000000   0.0000              0.0000           0.000   
31        M6G        0.000000   0.0000              0.0000           0.000   
32        M6H        0.000000   0.0000              0.0000           0.000   
33        M6J        0.000000   0.0000              0.0000           0.000   
34        M6K        0.000000   0.0000              0.0000           0.000   
35        M6P        0.000000   0.0000              0.0000           0.000   
36        M6R        0.000000   0.0000              0.0000           0.000   
37        M6S        0.000000   0.0000              0.0000           0.000   
38        M7Y        0.000000   0.0000              0.0000           0.000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0             0.000             0.000             0.000000      0.000000   
1             0.000             0.000             0.000000      0.000000   
2             0.000             0.000             0.023810      0.000000   
3             0.000             0.000             0.000000      0.000000   
4             0.000             0.000             0.055556      0.000000   
5             0.000             0.000             0.000000      0.000000   
6             0.000             0.000             0.000000      0.000000   
7             0.000             0.000             0.000000      0.000000   
8             0.000             0.000             0.000000      0.000000   
9             0.000             0.000             0.000000      0.000000   
10            0.000             0.000         

#### Let's print each Postal Area along with the top 5 most common venues


In [35]:
num_top_venues = 5

for hood in toronto_grouped['PostalCode']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['PostalCode'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M4E----
               venue  freq
0       Neighborhood  0.25
1  Health Food Store  0.25
2                Pub  0.25
3              Trail  0.25
4     Adult Boutique  0.00


----M4J----
                 venue  freq
0                 Park   0.5
1    Convenience Store   0.5
2  Monument / Landmark   0.0
3               Lounge   0.0
4     Malay Restaurant   0.0


----M4K----
                    venue  freq
0        Greek Restaurant  0.17
1             Coffee Shop  0.10
2      Italian Restaurant  0.07
3  Furniture / Home Store  0.05
4              Restaurant  0.05


----M4L----
                  venue  freq
0                  Park  0.13
1  Fast Food Restaurant  0.09
2     Food & Drink Shop  0.04
3                   Pub  0.04
4               Brewery  0.04


----M4M----
                 venue  freq
0          Coffee Shop  0.08
1              Brewery  0.06
2               Bakery  0.06
3            Gastropub  0.06
4  American Restaurant  0.06


----M4N----
            venue  freq
0           

#### Let's put that into a _pandas_ dataframe

First, let's write a function to sort the venues in descending order.


In [36]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 5 venues for each neighborhood.

In [38]:
import numpy as np

In [77]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_grouped['PostalCode']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

PostalCode 1st Most Common Venue 2nd Most Common Venue  \
0        M4E          Neighborhood                   Pub   
1        M4J                  Park     Convenience Store   
2        M4K      Greek Restaurant           Coffee Shop   
3        M4L                  Park  Fast Food Restaurant   
4        M4M           Coffee Shop                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0     Health Food Store                 Trail            Donut Shop  
1                 Diner        Farmers Market    Falafel Restaurant  
2    Italian Restaurant            Restaurant        Ice Cream Shop  
3               Brewery    Italian Restaurant     Fish & Chips Shop  
4                  Café   American Restaurant             Gastropub

Run _k_-means to cluster the neighborhood into 5 clusters.

In [78]:
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('PostalCode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 2, 0, 0, 0, 2, 0, 0, 0, 1])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [79]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df3

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('PostalCode'), on='PostalCode')

toronto_merged.head() # check the last columns!

PostalCode           Borough               Neighbourhood   Latitude  \
2         M5A  Downtown Toronto  Regent Park , Harbourfront  43.654260   
9         M5B  Downtown Toronto    Garden District, Ryerson  43.657162   
15        M5C  Downtown Toronto              St. James Town  43.651494   
19        M4E      East Toronto                 The Beaches  43.676357   
20        M5E  Downtown Toronto                 Berczy Park  43.644771   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
2  -79.360636               0           Coffee Shop                   Pub   
9  -79.378937               0           Coffee Shop        Clothing Store   
15 -79.375418               0           Coffee Shop                  Café   
19 -79.293031               0          Neighborhood                   Pub   
20 -79.373306               0           Coffee Shop          Cocktail Bar   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
2                   Park                Bakery        Breakfast Spot  
9                   Café                 Hotel        Sandwich Place  
15              Beer Bar          Cocktail Bar             Gastropub  
19     Health Food Store                 Trail            Donut Shop  
20                Bakery           Cheese Shop              Beer Bar

In [80]:
latitude,longitude = get_coordinates("Toronto City")

In [81]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['PostalCode'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<h4>Cluster 1</h4>

In [88]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[0,1]+ list(range(5, toronto_merged.shape[1]))]]

PostalCode                                            Borough  \
2          M5A                                   Downtown Toronto   
9          M5B                                   Downtown Toronto   
15         M5C                                   Downtown Toronto   
19         M4E                                       East Toronto   
20         M5E                                   Downtown Toronto   
24         M5G                                   Downtown Toronto   
25         M6G                                   Downtown Toronto   
30         M5H                                   Downtown Toronto   
31         M6H                                       West Toronto   
36         M5J                                   Downtown Toronto   
37         M6J                                       West Toronto   
41         M4K                                       East Toronto   
42         M5K                                   Downtown Toronto   
43         M6K                                       West Toronto   
47         M4L                                       East Toronto   
48         M5L                                   Downtown Toronto   
54         M4M                                       East Toronto   
62         M5N                                    Central Toronto   
67         M4P                                    Central Toronto   
69         M6P                                       West Toronto   
73         M4R                                    Central Toronto   
74         M5R                                    Central Toronto   
75         M6R                                       West Toronto   
79         M4S                                    Central Toronto   
80         M5S                                   Downtown Toronto   
81         M6S                                       West Toronto   
84         M5T                                   Downtown Toronto   
86         M4V                                    Central Toronto   
87         M5V                                   Downtown Toronto   
92         M5W     Downtown TorontoStn A PO Boxes25 The Esplanade   
96         M4X                                   Downtown Toronto   
97         M5X                                   Downtown Toronto   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   

     Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
2                 0           Coffee Shop                    Pub   
9                 0           Coffee Shop         Clothing Store   
15                0           Coffee Shop                   Café   
19                0          Neighborhood                    Pub   
20                0           Coffee Shop           Cocktail Bar   
24                0           Coffee Shop         Sandwich Place   
25                0         Grocery Store                   Café   
30                0           Coffee Shop                   Café   
31                0              Pharmacy                 Bakery   
36                0           Coffee Shop                   Café   
37                0                   Bar  Vietnamese Restaurant   
41                0      Greek Restaurant            Coffee Shop   
42                0           Coffee Shop                  Hotel   
43                0                  Café         Breakfast Spot   
47                0                  Park   Fast Food Restaurant   
48                0           Coffee Shop                  Hotel   
54                0           Coffee Shop                 Bakery   
62                0          Home Service            Music Venue   
67                0                  Park            Pizza Place   
69                0                  Café        Thai Restaurant   
73                0           Coffee Shop         Clothing Store   
74                0        Sandwich Place            Coffee Shop   
75                0     

<h4>Cluster 2</h4>

In [89]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode          Borough  Cluster Labels 1st Most Common Venue  \
83        M4T  Central Toronto               1                Lawyer   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83            Playground                 Diner        Farmers Market   

   5th Most Common Venue  
83    Falafel Restaurant

<h4>Cluster 3</h4>

In [90]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode                Borough  Cluster Labels 1st Most Common Venue  \
35        M4J  East YorkEast Toronto               2                  Park   
61        M4N        Central Toronto               2                  Park   
68        M5P        Central Toronto               2                  Park   
91        M4W       Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
35     Convenience Store                 Diner        Farmers Market   
61              Bus Line           Swim School            Comic Shop   
68      Sushi Restaurant         Jewelry Store                 Trail   
91                 Trail            Playground          Dessert Shop   

   5th Most Common Venue  
35    Falafel Restaurant  
61        Discount Store  
68    Falafel Restaurant  
91    Falafel Restaurant

We get 3 clusters with 2nd cluster having only one postal code of 'M4T' which seems different from all Postal areas as it do not have any Park or Restaurant or Cofee shops in its top 5 Venues. Thus it seems more formal or Non Residential area